File: samples/AdventureWorks/AdventureWorks.Client.Spa/DataObjects/Sales/CreditCardPaymentObjectCustomized.ts
import { CreditCardPaymentObject as GeneratedDataObject } from 'DataObjects/Sales/CreditCardPaymentObject';
import { PersonCreditCard } from 'Enumerations/Enumerations';
import { Header } from 'xomega';

export class CreditCardPaymentObject extends GeneratedDataObject {

    // perform post intialization
    onInitialized() {
        super.onInitialized();
        let obj = this;
        this.CreditCardId.InternalValue.subscribe(function (h: Header) {
            obj.CardNumber.InternalValue(h ? h.attr[PersonCreditCard.Attributes.CardNumber] : null);
            // for composite properties display null if an underlying property is null
            obj.Expiration.InternalValue(h && h.attr[PersonCreditCard.Attributes.ExpYear] ?
                `${h.attr[PersonCreditCard.Attributes.ExpMonth]}/${h.attr[PersonCreditCard.Attributes.ExpYear]}` : null);
        });
    }
}

File: demo/AdventureWorks.Client.Common/DataObjects/Sales/CreditCardPaymentObjectCustomized.cs
using System;
using AdventureWorks.Services.Common.Enumerations;
using Xomega.Framework;
using Xomega.Framework.Properties;

namespace AdventureWorks.Client.Common.DataObjects
{
    public class CreditCardPaymentObjectCustomized : CreditCardPaymentObject
    {
        public CreditCardPaymentObjectCustomized()
        {
        }

        public CreditCardPaymentObjectCustomized(IServiceProvider serviceProvider) : base(serviceProvider)
        {
        }

        // construct properties and child objects
        protected override void Initialize()
        {
            base.Initialize();
            // add custom construction code here
        }

        // perform post initialization
        protected override void OnInitialized()
        {
            base.OnInitialized();
            CreditCardIdProperty.Change += OnCreditCardChanged;
        }

        private void OnCreditCardChanged(object sender, PropertyChangeEventArgs e)
        {
            if (e.Change.IncludesValue() && !Equals(e.OldValue, e.NewValue))
            {
                Header cc = CreditCardIdProperty.Value;
                CardNumberProperty.SetValue(cc?[PersonCreditCard.Attributes.CardNumber]);
                ExpirationProperty.SetValue(cc == null ? null : cc[PersonCreditCard.Attributes.ExpMonth]
                                                        + "/" + cc[PersonCreditCard.Attributes.ExpYear]);
            }
        }
    }
}